### PR TITLE
Implement a heuristic for using fast-path serialization in streaming JsonSerializer methods.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR refactors the root-level serialization methods so that source generated fast-path serialization delegates can be utilized in more places, including `object` values and in the streaming serialization methods. Specifically for the streaming methods, a heuristic is implemented where fast-path serialization kicks in if the following conditions are satisfied for a given type:

1. No serialized value has generated a JSON payload exceeding `JsonSerializerOptions.DefaultBufferSize / 2` bytes AND
2. At least 10 serialization operations have already been completed.

This is to prevent excessive buffering since the fast path serializer doesn't support streaming.

## Performance

The change shows notable performance improvements in streaming serialization workloads:

### WriteJson_LoginViewModel

|            Method |        Job |                                                                                                          Toolchain |      Mode |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | MannWhitney(3%) | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|------------------ |----------- |------------------------------------------------------------------------------------------------------------------- |---------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|---------------- |--------:|-------:|----------:|------------:|
| SerializeToStream | Job-KZZOPZ |         main | SourceGen | 416.3 ns | 14.25 ns | 15.84 ns | 413.8 ns | 387.5 ns | 443.4 ns |  1.00 |            Base |    0.00 | 0.0141 |     152 B |        1.00 |
| SerializeToStream | Job-UEGCWD | PR | SourceGen | 344.4 ns | 32.70 ns | 37.65 ns | 345.7 ns | 294.5 ns | 436.7 ns |  0.83 |          Faster |    0.09 | 0.0150 |     152 B |        1.00 |

### WriteJson_LargeStructWithProperties

|                  Method |        Job |                                                                                                          Branch |      Mode |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | MannWhitney(3%) | RatioSD |   Gen0 | Allocated | Alloc Ratio |
|------------------------ |----------- |------------------------------------------------------------------------------------------------------------------- |---------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|---------------- |--------:|-------:|----------:|------------:|
|       SerializeToStream | Job-KZZOPZ |          main | SourceGen | 768.1 ns | 42.52 ns | 48.97 ns | 761.1 ns | 696.2 ns | 864.0 ns |  1.00 |            Base |    0.00 | 0.0211 |     232 B |        1.00 |
|       SerializeToStream | Job-UEGCWD | PR | SourceGen | 393.0 ns | 11.23 ns | 11.53 ns | 388.5 ns | 383.7 ns | 426.3 ns |  0.52 |          Faster |    0.02 | 0.0151 |     152 B |        0.66 |

### WriteJson_Dictionary

|            Method |        Job |                                                                                                          Toolchain |      Mode |      Mean |     Error |    StdDev |   Median |      Min |       Max | Ratio | MannWhitney(3%) | RatioSD | Allocated | Alloc Ratio |
|------------------ |----------- |------------------------------------------------------------------------------------------------------------------- |---------- |----------:|----------:|----------:|---------:|---------:|----------:|------:|---------------- |--------:|----------:|------------:|
| SerializeToStream | Job-KZZOPZ |          main | SourceGen | 10.260 μs | 1.3123 μs | 1.5112 μs | 9.767 μs | 8.691 μs | 13.380 μs |  1.00 |            Base |    0.00 |     152 B |        1.00 |
| SerializeToStream | Job-UEGCWD | PR | SourceGen |  7.558 μs | 0.6635 μs | 0.7640 μs | 7.557 μs | 6.435 μs |  9.027 μs |  0.75 |          Faster |    0.15 |     152 B |        1.00 |

### WriteJson_ImmutableDictionary

|            Method |        Job |                                                                                                          Toolchain |      Mode |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | MannWhitney(3%) | RatioSD | Allocated | Alloc Ratio |
|------------------ |----------- |------------------------------------------------------------------------------------------------------------------- |---------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|---------------- |--------:|----------:|------------:|
| SerializeToStream | Job-KZZOPZ |          main | SourceGen | 26.40 μs | 1.519 μs | 1.688 μs | 26.40 μs | 22.39 μs | 29.12 μs |  1.00 |            Base |    0.00 |     304 B |        1.00 |
| SerializeToStream | Job-UEGCWD | PR | SourceGen | 12.07 μs | 1.120 μs | 1.289 μs | 11.52 μs | 10.69 μs | 15.26 μs |  0.45 |          Faster |    0.05 |     152 B |        0.50 |

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
FYI @stephentoub @eerhardt @davidfowl

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Stream.WriteTests.cs [hunk 1]
```diff
@@ -5,31 +5,32 @@
...
     public partial class StreamTests
     {
         [Fact]
-        public static async Task WriteNullArgumentFail()
+        public async Task WriteNullArgumentFail()
         {
             await Assert.ThrowsAsync<ArgumentNullException>(async () => await JsonSerializer.SerializeAsync((Stream)null, 1));
             await Assert.ThrowsAsync<ArgumentNullException>(async () => await JsonSerializer.SerializeAsync((Stream)null, 1, typeof(int)));
-            Assert.Throws<ArgumentNullException>(() => JsonSerializer.Serialize((Stream)null));
+            Assert.Throws<ArgumentNullException>(() => JsonSerializer.Serialize((Stream)null, 1));
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Stream.WriteTests.cs:20]: can this be added as new assert?

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Stream.WriteTests.cs [hunk 2]
```diff
@@ -5,31 +5,32 @@
...
+using System.Text.Json.Serialization.Metadata;
 using Xunit;
 
 namespace System.Text.Json.Serialization.Tests
 {
     public partial class StreamTests
     {
         [Fact]
-        public static async Task WriteNullArgumentFail()
+        public async Task WriteNullArgumentFail()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Stream.WriteTests.cs:16]: Unrelated to the change itself, but these test weren't being run because they were marked as static.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs [hunk 3]
```diff
@@ -19,7 +18,7 @@ public static partial class JsonSerializer
         // We check for flush after each JSON property and element is written to the buffer.
         // Once the buffer is expanded to contain the largest single element\property, a 90% thresold
         // means the buffer may be expanded a maximum of 4 times: 1-(1/(2^4))==.9375.
-        private const float FlushThreshold = .90f;
+        internal const float FlushThreshold = .90f;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs:21]: I think you should move it to a new file - it doesn't seem to be used anywhere here

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs:21]: I figured I'd keep it here so that we can have git history on when/why it was added.

**Review by krwq — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    {
+        // This section provides helper methods guiding root-level serialization
+        // of values corresponding according to the current JsonTypeInfo configuration.
+
+        // Root serialization method for sync, non-streaming serialization
+        internal void Serialize(
+            Utf8JsonWriter writer,
+            in T? rootValue,
+            object? rootValueBoxed = null,
+            bool isInvokedByPolymorphicConverter = false)
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:23]: Can the logic that uses these values be refactored (yielding material benefit) into a separate private method since the only calls that pass them are within the type itself?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        }
+
+        // Fast-path serialization in source gen has not been designed with streaming in mind.
+        // Even though it's not used in streaming by default, we can sometimes try to turn it on
+        // assuming that the current type is known to produce small enough JSON payloads.
+        // The `CanUseSerializeHandlerInStreaming` flag returns true iff:
+        //  * The type has been used in at least `MinSerializationsSampleSize` streaming serializations AND
+        //  * No serialization size exceeding JsonSerializerOptions.DefaultBufferSize / 2 has been recorded so far.
+        private bool CanUseSerializeHandlerInStreaming => _canUseSerializeHandlerInStreamingState == 1;
+        private volatile int _canUseSerializeHandlerInStreamingState; // 0: unspecified, 1: allowed, 2: forbidden
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:303]: Can this be simplified into a simple yes/no decision? Otherwise can we use an enum for named constants?

**Review by layomia — COMMENTED** [review]
Minor questions but LGTM overrall.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs (hunk 5 above)
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:303]: not sure if compare exchange works with enums, if not you can use still use enum to assign but leave state as int (just cast it to int before assigning)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs [hunk 6]
```diff
@@ -1,127 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs:1]: We should do that eventually, but I felt doing it here would blow up scope beyond the goal of the PR.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs (hunk 4 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:23]: I considered this, but it would end up in a lot of duplication (4x duplication of serialization logic per mode (typed/untyped/polymorphic/non-polymorphic) x (Writer/Sync streaming/Async streaming)). I ended up with this compromise which doesn't sacrifice performance substantially.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs:1]: Pushed another commit that refactors the Read helpers for consistency.

**Comment by eiriktsarpalis** [comment]
We just merged https://github.com/dotnet/performance/pull/2740 adding source gen benchmarks to `dotnet/performance`. I'm holding off from merging this PR for a few days, to give our performance infrastructure the chance to properly record the performance numbers before this change.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+            if (CanUseSerializeHandlerInStreaming)
+            {
+                // Short-circuit calls into SerializeHandler, if the `CanUseSerializeHandlerInStreaming` heuristic allows it.
+
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(Options.SerializerContext?.CanUseSerializationLogic == true);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                using var writer = new Utf8JsonWriter(bufferWriter, Options.GetWriterOptions());
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:84]: Can this use `Utf8JsonWriterCache.RentWriterAndBuffer` instead of allocating a new `Utf8JsonWriter` every time?

Allocating a new Utf8JsonWriter seems to be one of the bigger allocs in ASP.NET - https://github.com/dotnet/aspnetcore/issues/45332.

In https://github.com/aspnet/Benchmarks/pull/1772/ we are manually caching the `Utf8JsonWriter` in the benchmark.

cc @DamianEdwards @BrennanConroy

Comment (BrennanConroy) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:84]: We can do that because we use the sync APIs right? Once you go async it'll mess up the current Utf8JsonWriterCache implementation I think.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:84]: Can we just reuse the `Utf8JsonWriter` and not the `PooledByteBufferWriter`? The `Utf8JsonWriter` doesn't seem to be used asynchronously here.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+
+                try
+                {
+                    SerializeHandler(writer, rootValue!);
+                    writer.Flush();
+                    await bufferWriter.WriteToStreamAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:94]: Yes. Moving serialization to the non-streaming fast path admits the possibility of it failing with OOM's, so we want to be able to invalidate the optimization as soon as it happens.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs (hunk 7 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs]: Yep, that can be done. I've updated the branch to include that change.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:93]: This comment is on the wrong line - it should be for the `else` case, which gets called the first 10 times.

My thinking is that if an exception is thrown, the full payload hasn't been written yet. If we call `OnRootLevelAsyncSerializationCompleted` with less than the full payload of bytes, it could count towards the heuristic, when it really shouldn't.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+        private T? UnboxValue(object? value)
+        {
+            if (
+#if NETCOREAPP
+                // Treated as a constant by recent versions of the JIT.
+                typeof(T).IsValueType &&
+#else
+                Type.IsValueType &&
+#endif
+                default(T) is not null && value is null)
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: I'm not sure I understand the point of the `default(T) is not null` check. If we know that `T` is a ValueType, when will `default(T) is not null` ever be `false`? It seems like a redundant check to me.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: Oh - it's for `Nullable<T>` ... right?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: It can only be true if your T is `Nullable<T>`, which is the case we're meant to catch here.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: FWIW this is pre-existing logic that got moved from `JsonConverterOfT.WriteCore.cs`

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: Hmm... actually you're right, that check should have been `default(T) is null && value is null`. Surprised this isn't caught by any tests.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: I _think_ the code is correct, now that I realized it was for checking for Nullable.  When T is Nullable, `typeof(T).IsValueType` will be `true` and `default(T) is not null` will be `false`. So the exception won't be thrown in the case of `value` being `null`. Because it is fine for `Nullable<T>` to have a `null` value.

I guess my one suggestion would be to throw a comment here to help the next person realize faster that this check is for Nullable.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:289]: > I guess my one suggestion would be to throw a comment here to help the next person realize faster that this check is for Nullable.

There is a comment about that in line 291.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+                bool isFinalBlock;
+                WriteStack state = default;
+                state.Initialize(this,
+                    rootValueBoxed,
+                    isInvokedByPolymorphicConverter,
+                    supportContinuation: true,
+                    supportAsync: true);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                using var writer = new Utf8JsonWriter(bufferWriter, Options.GetWriterOptions());
```
Comment (BrennanConroy) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:122]: Has any thought been put into making the cache use a `ConcurrentQueue` instead so it can be used in async paths?
Example: https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: Given that the perf benefits are [fairly modest](https://github.com/dotnet/runtime/pull/73338) when doing thread-local caching in the sync methods, my expectation is that any mechanism requiring more elaborate synchronization would likely offer diminishing returns.

Comment (DamianEdwards) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: The big gain we've seen in our JSON web benchmarks is that by reducing the allocations of the `Utf8JsonWriter` (aka pooling them) it greatly reduces the working set of the application, from over 400 MB to ~60 MB for a test run doing millions of RPS.

Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: > The big gain we've seen in our JSON web benchmarks is that by reducing the allocations of the `Utf8JsonWriter` (aka pooling them) it greatly reduces the working set of the application, from over 400 MB to ~60 MB for a test run doing millions of RPS.

Do we know how much of that would also be achieved by changing GC configuration to make it more aggressive rather than being ok expanding to the memory available?

Comment (DamianEdwards) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: @BrennanConroy could we do a run of the JSON middleware benchmark with Workstation GC to compare?

Comment (BrennanConroy) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: Server GC: 468MB working set
Workstation GC: 99MB working set

Server GC with manual caching: 86MB working set
Workstation GC with manual caching: 84MB working set

Comment (DamianEdwards) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: Thanks @BrennanConroy. For the sake of curiosity, what were the RPS numbers for those runs?

The savings when pooling occurs seem worthwhile even when GC is more aggressive (15% for our JSON middleware benchmark).

Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: @Maoni0, if someone wants the scalability of server GC but also wants to prioritize minimizing working set, what are the right knobs to experiment with?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: None of the streaming methods accept `Utf8JsonWriter`, so I'm guessing these numbers came from benchmarking the [non-streaming method](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serialize?view=net-7.0#system-text-json-jsonserializer-serialize-1(system-text-json-utf8jsonwriter-0-system-text-json-jsonserializeroptions)) to itself using caching? If not, I don't think it would be an apples-to-apples comparison, particularly from an allocation perspective.

Comment (BrennanConroy) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: > I'm guessing these numbers came from benchmarking the [non-streaming method](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serialize?view=net-7.0#system-text-json-jsonserializer-serialize-1(system-text-json-utf8jsonwriter-0-system-text-json-jsonserializeroptions)) to itself using caching?

Correct.

I also realized the numbers I shared before were wrong since we switched to using the source-generator at the same time.

So here are the numbers with just caching Utf8Writer, no source-gen or JsonConverter to avoid the struct boxing so there is ~24M bytes/sec allocations:

Server GC: 468MB working set
Workstation GC: 99MB working set

Server GC with manual caching: 427MB working set
Workstation GC with manual caching: 98MB working set

Basically, working set with server GC will be high if there are any allocations it looks like.

Comment (davidfowl) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: This is maybe the only thing similar right now in the JSON code base https://github.com/dotnet/runtime/blob/b77aa8a9726503df52327a0388a3f4a0325989e1/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs#L166

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: Could you share the benchmark code? If run in .NET 7, I would expect any perf issues related to `Utf8JsonWriter` allocations to have dissipated because of #73338.

Comment (BrennanConroy) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: https://github.com/aspnet/Benchmarks/blob/main/src/Benchmarks/Middleware/JsonMiddleware.cs#L56
It's using the Stream methods which weren't affected until this PR.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: What about the variant that caches `Utf8JsonWriter` instances?

Comment (BrennanConroy) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs:125]: Still streaming API, just the synchronous one
https://github.com/aspnet/Benchmarks/pull/1772
